stlab/forest_algorithms.hpp

```diff
@@ -0,0 +1,122 @@
+/**************************************************************************************************/
+
+#ifndef STLAB_FOREST_ALGORITHMS_HPP
+#define STLAB_FOREST_ALGORITHMS_HPP
+
+/**************************************************************************************************/
+
+// stdc++
+#include <optional>
+
+// stlab
+#include <stlab/forest.hpp>
+
+/**************************************************************************************************/
+
+namespace stlab {
+
+/**************************************************************************************************/
+// "Congruent" would be a nice name here, but in geometry that also implies reflection.
+template <class Forest1, class Forest2>
+bool equal_shape(const Forest1& x, const Forest2& y) {
+    if (x.size_valid() && y.size_valid() && x.size() != y.size()) return false;
+    auto pos{y.begin()};
+    for (auto first(x.begin()), last(x.end()); first != last; ++first, ++pos) {
+        if (first.edge() != pos.edge()) return false;
+    }
+    return true;
+}
+
+/**************************************************************************************************/
+
+template <class Forest>
+struct forest_insert_iterator {
+    using value_type = typename Forest::value_type;
+    using iterator_type = typename Forest::iterator;
+
+    explicit forest_insert_iterator(Forest& f) : _f{f}, _p{_f.root()} {}
+
+    auto& operator*() { return *this; }
+
+    auto& operator++() {
+        ++_p;
+        return *this;
+    }
+
+    auto& operator=(value_type&& x) {
+        _p = _f.insert(_p, std::forward<value_type>(x));
+        return *this;
+    }
+
+    void trailing() { _p = trailing_of(_p); }
+
+private:
+    Forest& _f;
+    iterator_type _p;
+};
+
+template <class Forest>
+auto forest_inserter(Forest& f) {
+    return forest_insert_iterator(f);
+}
+
+/**************************************************************************************************/
+
+template <class I, class O, class P, class UP>
+auto transform_forest(I first, I last, O out, P&& proj, UP&& pred) {
+    for (; first != last; ++first) {
+        ++out;
+        if (pred(first)) {
+            *out = proj(*first);
+        } else {
+            out.trailing();
+        }
+    }
+    return out;
+}
+
+/**************************************************************************************************/
+
+template <class Forest,
+          class P,
+          class U = decltype(std::declval<P>()(typename Forest::value_type()))>
+auto transcribe_forest(const Forest& f, P&& proj) {
+    typename Forest::template rebind<U>::type result;
+    transform_forest(f.begin(), f.end(), forest_inserter(result), std::forward<P>(proj),
```

Comment:
Does forest implements cbegin() and cend(), if so , I would use them here.

Reply:
It does not - it has `const` variants of `begin` and `end` that return `const_iterator`s. I'll replace them with calls to `std::cbegin` and `std::cend`, though, to be explicit.